### ISSUE TITLE
With blaze 304 (Not Modified) gets a body although it shouldn't

### ISSUE THREAD

**Issue by hvesalai** [http4s/http4s#549]
We have code that writes

```Response(status = Status.NotModified).withBody("Not modified")```

In `Status.scala` `NotModified` has been defined with `isEntityAllowed = false`

Yet http4s writes the body to the response, breaking the HTTP specification (and causing issues with persistent connections).

**Comment by hvesalai** [comment]
I thought `Http1ServerStage.scala` `renderResponse` should prevent this (by using the BodylessWriter), but apparently it doesn't. So I'm totally lost on why this happens.

We are using version 0.12.1

**Comment by hvesalai** [comment]
Ok. Now I do understand. The complete test in `Http1ServerState` is:

```
(!resp.status.isEntityAllowed && lengthHeader.isEmpty && respTransferCoding.isEmpty)
```

Meaning if the poor API user give a body to 304, then http4s blaze will send it to the line. I have a feeling we discussed this and somebody suggested back then that http4s should allow the user to shoot his leg. I would like to again propose that http4s should not allow the user to shoot his leg this way.

**Comment by hvesalai** [comment]
Curl output of the service:

```
< HTTP/1.1 304 Not Modified
< Content-Length: 12
< Content-Type: text/plain; charset=UTF-8
< Date: Mon, 07 Mar 2016 12:19:53 UTC
< 
* Excess found in a non pipelined read: excess = 12 url = /portfolio/statistics?portfolioId=0&h=-444974834-1498970459--771277118 (zero-length body)
* Connection #0 to host localhost left intact
```

**Comment by hvesalai** [comment]
Furthermore `The response MUST NOT include other entity-headers; this prevents inconsistencies between cached entity-bodies and updated headers. `

So not only is it wrong that the response had the body, it shouldn't even have the `Content-Length` header (not even with a value of 0).

**Comment by bryce-anderson** [comment]
I was for 'letting the user shot themselves in the leg'. Is there a reason that you're attaching a body to a status code that doesn't allow it especially when you don't want it?

I'm open to changing this behavior. The first problem is pretty simple: just ignore the body. (I believe that is what we are doing, not killing it, right?). I've not looked into what it would take to not send entity-headers.

**Comment by bryce-anderson** [comment]
@hvesalai, if you decided to put a PR up for this we have a branch `release-0.12.4` that is getting primed for launch.

**Comment by hvesalai** [comment]
I'm on it.

**Comment by bryce-anderson** [comment]
Closed via #551.
